python/pyarrow/tests/test_dataset.py

```diff
@@ -622,28 +622,113 @@ def test_make_fragment(multisourcefs):
+    assert len(result) == 4
+
+    # scanning fragments follow filter predicate
+    fragments = list(dataset.get_fragments(filter=ds.field('b') < 2))
+    assert len(fragments) == 2
+    result = fragments[0].to_table()
+    assert result.column_names == ['b', 'c', 'a']
+    assert len(result) == 2
+    result = fragments[1].to_table()
+    assert len(result) == 0
+
+    # manually re-construct a fragment
+    fragment = list(dataset.get_fragments())[0]
+    parquet_format = fragment.format
+    new_fragment = parquet_format.make_fragment(
+        fragment.path, fragment.filesystem, schema=dataset.schema,
+        partition_expression=fragment.partition_expression)
+    assert new_fragment.to_table().equals(fragment.to_table())
+    new_fragment.to_table()
+
+    # manually re-construct a fragment with filter / column projection
+    new_fragment = parquet_format.make_fragment(
+        fragment.path, fragment.filesystem,  # schema=dataset.schema,
+        columns=['b'], filter=ds.field('b') < 2,
+        partition_expression=fragment.partition_expression)
+    result = new_fragment.to_table()
+    assert result.column_names == ['b']
+    assert len(result) == 2
+
+    # list and scan row group fragments
+    fragment = list(dataset.get_fragments())[0]
+    row_group_fragments = list(fragment.get_row_group_fragments())
+    assert len(row_group_fragments) == 2
+    result = row_group_fragments[0].to_table()
+    assert result.column_names == ['b', 'c', 'a']
+    assert len(result) == 2
+
+    # scanning row group fragment follows column projection / filter predicate
+    fragment = list(dataset.get_fragments(
+        columns=['a', 'b'], filter=ds.field('b') < 1))[0]
+    row_group_fragments = list(fragment.get_row_group_fragments())
+    assert len(row_group_fragments) == 1
+    result = row_group_fragments[0].to_table()
+    assert result.column_names == ['a', 'b']
+    assert len(result) == 1
+
+    # manually re-construct row group fragments
+    fragment = list(dataset.get_fragments())[0]
+    row_group_fragments = list(fragment.get_row_group_fragments())
+    new_fragment = parquet_format.make_fragment(
+        fragment.path, fragment.filesystem,  # schema=dataset.schema,
+        partition_expression=fragment.partition_expression,
+        row_groups=[0])
+    result = new_fragment.to_table()
+    # TODO needs to include partition column
+    # assert result.equals(row_group_fragments[0].to_table())
+    assert result.column_names == ['b', 'c']
+    assert len(result) == 2
+
+    # manually re-construct a row group fragment with filter/column projection
+    new_fragment = parquet_format.make_fragment(
+        fragment.path, fragment.filesystem,  # schema=dataset.schema,
+        columns=['b'], filter=ds.field('b') < 3,
+        partition_expression=fragment.partition_expression,
         row_groups={1})
+    result = new_fragment.to_table()
+    assert result.column_names == ['b']
+    assert len(result) == 1
```

Comment:
```suggestion
    new_fragment = parquet_format.make_fragment(
        fragment.path, fragment.filesystem, schema=dataset.schema,
        columns=['a', 'b'], filter=ds.field('b') < 3,
        partition_expression=fragment.partition_expression,
        row_groups={1})
    result = new_fragment.to_table()
    assert result.column_names == ['a', 'b']
    assert len(result) == 1
```

(to ensure we can also select the partition column here)